# Transfer Resolution Models

The following models take into account that, throughout analyses, *no_goal* seems to behave entirely differently to the other two Goal Type conditions (*goal_frequent* and *goal_non_frequent*). 

Previous focus structure effects (see notebooks 11 to 6) as well as furhter exhaustive modelling (see notebooks 13_appendix vs. 14_appendix) show consistently the heterogeneity of combined data; by spliting data into goal contexts and no-goal contexts, both effects as well as models improved despite the pruning of observations. No-goal contexts are also analysed separately; however, and although results are coherent with what is seen by comparing combined data to goal behaviour data, the sample is clearly not enough to show anything significant from no-goal contexts.

Taking into account the effects summary (notebook 5) and response opportunity analysis (notebook 6), these were chosen to be the best models to predict participants resolving by means of transfer.

Import Libraries

In [39]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Goal Behaviour Subgroup

1. Create resolved_transfer (goals and no_goals)
2. Counts and Proportions (goals and no_goals)
3. Logistic Models

    - Goal Contexts

        Main Effect Model:

        - RT01: Focus

        Additive Model:

        - RT05: Focus + Goal_Type
        - RT07: Focus + Agent + Goal_Type

        Interactive Model:

        - RT11: Focus * Agent + Goal_Type

    - No-goal Contexts

        Main Effect Models:

        - RT12: Focus
        - RT13: Agent

        Additive Model:

        - RT14: Focus + Agent

        Interactive Model:

        - RT15: Focus * Agent

4. Model Comparisons
5. Summary

Read Data

In [40]:
subgroup_theoretical = pd.read_csv("../../data/processed/subgroup_theoretical.csv")

1. Create resolved_transfer

Binary Outcome: transfer responses from those observations that managed to escape L2_other options.

In [41]:
subgroup_theoretical["escape_L2"] = (
    subgroup_theoretical["Response_Full"] != "L2_other"
).astype(int)

In [42]:
escapees = subgroup_theoretical[subgroup_theoretical["escape_L2"] == 1].copy()

In [43]:
escapees["resolved_transfer"] = (escapees["Response_Full"] == "L1_transfer").astype(int)

Split escapees data into goals vs. no_goals

In [44]:
goals_escapees = escapees[escapees["Goal_Type"] != "no_goal"].copy()

In [45]:
no_goal_escapees = escapees[escapees["Goal_Type"] == "no_goal"].copy()

2. Counts and Proportions for goals and no_goals

General resolved_correct reponses count:

In [46]:
resolved_transfer_counts = pd.crosstab(
    escapees["resolved_transfer"],
    escapees["Response_Full"],
    margins=True)

resolved_transfer_counts

Response_Full,L1_transfer,correct,missing_response,All
resolved_transfer,,,,
0,0,61,3,64
1,96,0,0,96
All,96,61,3,160


resolved_transfer responses count for goals:

In [47]:
goals_resolved_transfer_counts = pd.crosstab(
    goals_escapees["resolved_transfer"],
    goals_escapees["Response_Full"],
    margins=True)

goals_resolved_transfer_counts

Response_Full,L1_transfer,correct,missing_response,All
resolved_transfer,,,,
0,0,43,1,44
1,62,0,0,62
All,62,43,1,106


resolved_transfer responses count for no_goals:

In [48]:
no_goal_resolved_transfer_counts = pd.crosstab(
    no_goal_escapees["resolved_transfer"],
    no_goal_escapees["Response_Full"],
    margins=True)

no_goal_resolved_transfer_counts

Response_Full,L1_transfer,correct,missing_response,All
resolved_transfer,,,,
0,0,18,2,20
1,34,0,0,34
All,34,18,2,54


General resolved_transfer responses proportions:

In [49]:
resolved_transfer_props = escapees["resolved_transfer"].value_counts(normalize = True)
resolved_transfer_props

resolved_transfer
1    0.6
0    0.4
Name: proportion, dtype: float64

From 160 observations, 40% of responses that had escaped *L2_other* options were resolved by means of transfer. 

resolved_correct responses proportions for goals:

In [50]:
goals_resolved_transfer_props = goals_escapees["resolved_transfer"].value_counts(normalize=True)
goals_resolved_transfer_props

resolved_transfer
1    0.584906
0    0.415094
Name: proportion, dtype: float64

From 106 observations, 41.5% of goal responses that had escaped *L2_other* options were resolved by means of transfer. Slightly higher than resolving correctly (40.5%) but not that much. 

resolved_correct responses proportions for no_goals:

In [51]:
no_goal_resolved_transfer_props = no_goal_escapees["resolved_transfer"].value_counts(normalize=True)
no_goal_resolved_transfer_props

resolved_transfer
1    0.62963
0    0.37037
Name: proportion, dtype: float64

From 54 observations, 37% of no_goal responses that had escaped *L2_other* options were resolved by means of transfer. Also higher than resolving correctly (33.3%) but not that much either.

- The proportion distribution of transfer resolution responses is higher when there is a goal oriented representation (41.5%) as opposed to when there isn't a representation of a goal (37%). 
- However, the gap narrows compared to the distribution of correct resolution responses between goal oriented vs. no goal representations.

Counts and Proportions by Condition:

In [52]:
counts_condition = pd.crosstab(
    [escapees["Goal_Type"], escapees["Agent"], escapees["Focus"]],
    escapees["resolved_transfer"],
    margins = True)

counts_condition

resolved_transfer               0   1  All
Goal_Type         Agent Focus             
goal_frequent     0     I      10   6   16
                        They    5   7   12
                  1     I       7   7   14
                        They    3   5    8
goal_non_frequent 0     I       9   7   16
                        They    3  14   17
                  1     I       4  11   15
                        They    3   5    8
no_goal           0     I       8  13   21
                        They    3  10   13
                  1     I       5   9   14
                        They    4   2    6
All                            64  96  160

In [53]:
goals_counts_condition = pd.crosstab(
    [goals_escapees["Goal_Type"], goals_escapees["Agent"], goals_escapees["Focus"]],
    goals_escapees["resolved_transfer"],
    margins = True)

goals_counts_condition

resolved_transfer               0   1  All
Goal_Type         Agent Focus             
goal_frequent     0     I      10   6   16
                        They    5   7   12
                  1     I       7   7   14
                        They    3   5    8
goal_non_frequent 0     I       9   7   16
                        They    3  14   17
                  1     I       4  11   15
                        They    3   5    8
All                            44  62  106

In [54]:
no_goal_counts_condition = pd.crosstab(
    [no_goal_escapees["Goal_Type"], no_goal_escapees["Agent"], no_goal_escapees["Focus"]],
    no_goal_escapees["resolved_transfer"],
    margins = True)

no_goal_counts_condition

resolved_transfer       0   1  All
Goal_Type Agent Focus             
no_goal   0     I       8  13   21
                They    3  10   13
          1     I       5   9   14
                They    4   2    6
All                    20  34   54

In [55]:
goals_props_condition = pd.crosstab(
    [goals_escapees["Goal_Type"], goals_escapees["Agent"]],
    goals_escapees["resolved_transfer"],
    normalize= "index")

goals_props_condition

resolved_transfer               0         1
Goal_Type         Agent                    
goal_frequent     0      0.535714  0.464286
                  1      0.454545  0.545455
goal_non_frequent 0      0.363636  0.636364
                  1      0.304348  0.695652

In [56]:
no_goal_props_condition = pd.crosstab(
    [no_goal_escapees["Goal_Type"], no_goal_escapees["Agent"]],
    no_goal_escapees["resolved_transfer"],
    normalize= "index")

no_goal_props_condition

resolved_transfer         0         1
Goal_Type Agent                      
no_goal   0        0.323529  0.676471
          1        0.450000  0.550000

3. Logistic Models

Import Libraries

In [57]:
import statsmodels.formula.api as smf
from scipy.stats import chi2

Sanity checks:

In [58]:
escapees["resolved_transfer"].value_counts()

resolved_transfer
1    96
0    64
Name: count, dtype: int64

In [59]:
goals_escapees["resolved_transfer"].value_counts()

resolved_transfer
1    62
0    44
Name: count, dtype: int64

In [60]:
no_goal_escapees["resolved_transfer"].value_counts()

resolved_transfer
1    34
0    20
Name: count, dtype: int64

**Goal Context**

**Main Effect Model**

**RT01**

resolved_transfer ~ Focus

In [61]:
RT01 = smf.logit(
    "resolved_transfer ~ Focus",
    data=goals_escapees
    ).fit()

print(RT01.summary())

Optimization terminated successfully.
         Current function value: 0.662011
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                  106
Model:                          Logit   Df Residuals:                      104
Method:                           MLE   Df Model:                            1
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                 0.02453
Time:                        14:15:52   Log-Likelihood:                -70.173
converged:                       True   LL-Null:                       -71.938
Covariance Type:            nonrobust   LLR p-value:                   0.06029
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept         0.0328      0.256      0.128      0.898      -0.469       0.535
Focus[T.They]     0.

With goal split dataset, the effect of Focus on its own seems to be more defined, almost reaching significance.

**Additive Models**

**RT05**

resolved_transfer ~ Focus + Goal_Type

In [62]:
RT05 = smf.logit(
    "resolved_transfer ~ Focus + Goal_Type",
    data=goals_escapees
    ).fit()

print(RT05.summary())

Optimization terminated successfully.
         Current function value: 0.649658
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                  106
Model:                          Logit   Df Residuals:                      103
Method:                           MLE   Df Model:                            2
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                 0.04273
Time:                        14:15:52   Log-Likelihood:                -68.864
converged:                       True   LL-Null:                       -71.938
Covariance Type:            nonrobust   LLR p-value:                   0.04623
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                         -0.2983      0.331     -0.901     

Even though the coefficients do not make it to significance, the model is significant overall. This is interesting as it does give way to consider that, despite not having enough power, Goal Type could after all play a role. 

**RT07**

resolved_transfer ~ Focus + Agent + Goal_Type

In [63]:
RT07= smf.logit(
    "resolved_transfer ~ Focus + Agent + Goal_Type",
    data=goals_escapees
    ).fit()

print(RT07.summary())

Optimization terminated successfully.
         Current function value: 0.645294
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                  106
Model:                          Logit   Df Residuals:                      102
Method:                           MLE   Df Model:                            3
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                 0.04916
Time:                        14:15:52   Log-Likelihood:                -68.401
converged:                       True   LL-Null:                       -71.938
Covariance Type:            nonrobust   LLR p-value:                   0.06960
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                         -0.4952      0.393     -1.262     

The full additive model, unlike for escaping the L2, is not significant. Yet, both the Focus as well as the goal type coefficients have improved once Agent in in the model. 

**Interactive Model**

**RT11**

resolved_transfer ~ Focus * Agent + Goal_Type

In [64]:
RT11 = smf.logit(
    "resolved_transfer ~ Focus * Agent + Goal_Type",
    data=goals_escapees
    ).fit()

print(RT11.summary())

Optimization terminated successfully.
         Current function value: 0.634611
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                  106
Model:                          Logit   Df Residuals:                      101
Method:                           MLE   Df Model:                            4
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                 0.06490
Time:                        14:15:52   Log-Likelihood:                -67.269
converged:                       True   LL-Null:                       -71.938
Covariance Type:            nonrobust   LLR p-value:                   0.05318
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                         -0.7155      0.425     -1.682     

The Focus' coefficient is significant in the interactive model. The Agent's coefficient, while not significant, has improved a lot comapared to the full additive model confirming that the relationship is interactive even if the interaction's coefficient is not.  
Goal type's contribution remains very steady, with a very similar contribution (coeff = 0.6518) compared to the additive model Focus-Goal_Type RT05 (coeff = 0.6531) and p-value = 0.115 (vs. 0.108). 

**No-goal Context**

**Main Effect Models**

**RT12**

resolved_transfer ~ Focus

In [65]:
RT12= smf.logit(
    "resolved_transfer ~ Focus",
    data=no_goal_escapees
    ).fit()

print(RT12.summary())

Optimization terminated successfully.
         Current function value: 0.659148
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                   54
Model:                          Logit   Df Residuals:                       52
Method:                           MLE   Df Model:                            1
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:               6.712e-06
Time:                        14:15:52   Log-Likelihood:                -35.594
converged:                       True   LL-Null:                       -35.594
Covariance Type:            nonrobust   LLR p-value:                    0.9826
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept         0.5261      0.350      1.504      0.133      -0.160       1.212
Focus[T.They]     0.

**RT13**

resolved_transfer ~ Agent

In [66]:
RT13= smf.logit(
    "resolved_transfer ~ Agent",
    data=no_goal_escapees
    ).fit()

print(RT13.summary())

Optimization terminated successfully.
         Current function value: 0.651219
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                   54
Model:                          Logit   Df Residuals:                       52
Method:                           MLE   Df Model:                            1
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                 0.01204
Time:                        14:15:52   Log-Likelihood:                -35.166
converged:                       True   LL-Null:                       -35.594
Covariance Type:            nonrobust   LLR p-value:                    0.3546
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.7376      0.367      2.012      0.044       0.019       1.456
Agent         -0.5369      0.

**Interactive Model**

**RT15**

resolved_transfer ~ Focus * Agent

In [67]:
RT15= smf.logit(
    "resolved_transfer ~ Focus * Agent",
    data=no_goal_escapees
    ).fit()

print(RT15.summary())

Optimization terminated successfully.
         Current function value: 0.628175
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                   54
Model:                          Logit   Df Residuals:                       50
Method:                           MLE   Df Model:                            3
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                 0.04700
Time:                        14:15:52   Log-Likelihood:                -33.921
converged:                       True   LL-Null:                       -35.594
Covariance Type:            nonrobust   LLR p-value:                    0.3413
                          coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------
Intercept               0.4855      0.449      1.080      0.280      -0.395       1.366
Fo

As expected, once no_goals are analysed on their own, 54 observations are truly not enough to see anything. 

4. Model Comparisons

In [68]:
from scipy.stats import chi2

def lr_test(model_small, model_large):

    LR = 2 * (model_large.llf - model_small.llf)

    df = model_large.df_model - model_small.df_model

    p = chi2.sf(LR, df)

    return pd.Series({
        "LR": LR,
        "df": df,
        "p_value": p
    })

**Unique contributions to the additive model**

Focus (RT01)

↓

Focus + Goal_Type (RT05)

In [69]:
lr_test(RT01, RT05)

LR         2.618956
df         1.000000
p_value    0.105594
dtype: float64

- Does Goal_Type add information after Focus?

Not significantly.

Focus + Goal_Type (RT05)

↓

Focus + Agent + Goal_Type (RT07)

In [70]:
lr_test(RT05, RT07)

LR         0.925055
df         1.000000
p_value    0.336151
dtype: float64

- Does Agency add information after Focus and Goal_Type?

As expected, Agency doesn't add information after Goal_Type and Focus; but it also shows how Goal Type, given the previous model comparison, is more relevant than Agency.

**Interaction contributions**

Focus + Agent + Goal_Type (RT07)

↓

Focus * Agent + Goal_Type (RT11)

In [71]:
lr_test(RT07, RT11)

LR         2.264766
df         1.000000
p_value    0.132346
dtype: float64

- Does the key Focus-Agent interaction still improve fit after Goal_Type is controlled?

The interaction does not survive; but given the previous model comparison, it does show that there would be room for an interaction if the sample was big enough.

**No_Goal contribution**

Agent (RT13)

↓

Focus * Agent (RT15)

In [72]:
lr_test(RT13, RT15)

LR         2.488782
df         2.000000
p_value    0.288116
dtype: float64

- Does the relationship require an interaction?

No, the relationship does not require an interaction. 

5. Summary 

From models, overall, what we see is:

- Escape:
    - Goal contexts: Agency dominates
    - No-goal contexts: Focus dominates
   
- Transfer Resolution:
    - Goal contexts: Focus dominates
    - No-goal contexts: Agency is better than Focus but the interaction may signal complexity.

Goal Behaviour could be acting as a latent behavioural regime rather than just another predictor. It seems that pursuing a goal could determine which cognitive variable governs counterfactual processing and desiderative extension. Participants who pursue goals resolve counterfactuals primarily through perspective (Focus), whereas participants without active goals resolve them primarily through responsibility (Agency). Escape behaviours display the complementary pattern.


From model comparisons:

- Focus is the only predictor that keeps sitting on the edge of significance after Goal is controlled for transfer resolution once the data focuses on goal behaviour.
